[MESOS][SPARK-8873] Clean up external shuffle data on driver exit with Mesos.

tnachen opened the pull request:
@andrewor14 @dragos

tnachen committed:
Clean up external shuffle data on driver exit with Mesos.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39186 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39186/consoleFull) for   PR 7820 at commit [`fb11a33`](https://github.com/apache/spark/commit/fb11a3313e23ea4e7ed1fb1ee82644ef2ad49380).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39188 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39188/consoleFull) for   PR 7820 at commit [`1910bd5`](https://github.com/apache/spark/commit/1910bd526c5378ea2a6fb5ff008751ed544a24d3).

SparkQA commented:
[Test build #39186 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39186/console) for   PR 7820 at commit [`fb11a33`](https://github.com/apache/spark/commit/fb11a3313e23ea4e7ed1fb1ee82644ef2ad49380).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #39188 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39188/console) for   PR 7820 at commit [`1910bd5`](https://github.com/apache/spark/commit/1910bd526c5378ea2a6fb5ff008751ed544a24d3).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class RegisterMesosDriver(appId: String, address: RpcAddress)`
  * `class MulticlassClassificationEvaluator (override val uid: String)`
  * `class NaiveBayes(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol):`
  * `class NaiveBayesModel(JavaModel):`
  * `class MulticlassClassificationEvaluator(JavaEvaluator, HasLabelCol, HasPredictionCol):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39281 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39281/consoleFull) for   PR 7820 at commit [`1910bd5`](https://github.com/apache/spark/commit/1910bd526c5378ea2a6fb5ff008751ed544a24d3).

SparkQA commented:
[Test build #39281 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39281/console) for   PR 7820 at commit [`1910bd5`](https://github.com/apache/spark/commit/1910bd526c5378ea2a6fb5ff008751ed544a24d3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class RegisterMesosDriver(appId: String, address: RpcAddress)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
retest this please

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,127 @@
+  }
+}
+
+case class RegisterMesosDriver(appId: String, address: RpcAddress)
```
private

andrewor14 replied:
also please add a simple java doc, even something like `A message sent from the driver to register with this shuffle service.` would do.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,127 @@
+
+  override def start(): Unit = {
+    super.start()
+    val port = conf.getInt("spark.mesos.shuffle.service.port", 7327)
```
no need to make a new config for this one, just reuse `spark.shuffle.service.port` at 7337

tnachen replied:
It won't clash on the same port?

andrewor14 replied:
why would you have both shuffle services on the same cluster? Your new shuffle service is a replacement for the old one, no?

tnachen replied:
It's actually a additional endpoint on a different port that is listening for the driver.

Let me take a look if I can easily manage to use the same, but it looks like it's not using the endpoint mechanism.

> On Aug 1, 2015, at 11:59 AM, andrewor14 <notifications@github.com> wrote:
> 
> In core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
> 
> > +
> > +import scala.collection.mutable
> > +
> > +import org.apache.spark.deploy.ExternalShuffleService
> > +import org.apache.spark.rpc.{RpcAddress, RpcEnv, ThreadSafeRpcEndpoint}
> > +import org.apache.spark.util.Utils
> > +import org.apache.spark.{Logging, SecurityManager, SparkConf}
> > +
> > +private[mesos] class MesosExternalShuffleService(conf: SparkConf, securityManager: SecurityManager)
> > +  extends ExternalShuffleService(conf, securityManager) {
> > +
> > +  private var rpcEnv: RpcEnv = _
> > +
> > +  override def start(): Unit = {
> > +    super.start()
> > +    val port = conf.getInt("spark.mesos.shuffle.service.port", 7327)
> why would you have both shuffle services on the same cluster? Your new shuffle service is a replacement for the old one, no?
> 
> —
> Reply to this email directly or view it on GitHub.
>

andrewor14 replied:
oh I see, currently 7337 is for executors while 7327 is for the driver. Yeah I think it would be very valuable to avoid another config.

andrewor14 replied:
One thing you could do is have a `MesosExternalShuffleBlockHandler` that extends `ExternalShuffleBlockHandler`. It would override the `receive` method to also accept a `RegisterDriver` message. Note that because it extends `RpcHandler` there is also a `connectionTerminated` method, so we can try using that one instead.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,127 @@
+import org.apache.spark.util.Utils
+import org.apache.spark.{Logging, SecurityManager, SparkConf}
+
+private[mesos] class MesosExternalShuffleService(conf: SparkConf, securityManager: SecurityManager)
```
This is missing a top level javadoc. You need to explain why Mesos needs a special shuffle service, i.e. it needs to detect when drivers exit, and the way we do it is through a separate RPC channel between the driver and each shuffle service.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,127 @@
+    private val connectedApps = new mutable.HashMap[RpcAddress, String]
+
+    override def receive: PartialFunction[Any, Unit] = {
+      case RegisterMesosDriver(appId, address) =>
```
This isn't a Mesos driver right? This is a Spark driver. I would just call this `RegisterDriver`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,127 @@
+    super.stop()
+  }
+
+  private class MesosExternalShuffleServiceEndpoint(override val rpcEnv: RpcEnv)
```
can you add a java doc for this class:
```
An RPC endpoint for talking to the Spark driver.
```

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,127 @@
+          }
+        } else {
+          connectedApps(address) = appId
+        }
```
should this be an else case? In L68 we removed the existing application ID if there's a duplicate. Shouldn't we put the updated one in `connectedApps`?

SparkQA commented:
[Test build #39344 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39344/consoleFull) for   PR 7820 at commit [`1910bd5`](https://github.com/apache/spark/commit/1910bd526c5378ea2a6fb5ff008751ed544a24d3).

andrewor14 commented on core/src/main/scala/org/apache/spark/rpc/RpcEndpoint.scala:
```diff
@@ -44,14 +44,14 @@ private[spark] trait ThreadSafeRpcEndpoint extends RpcEndpoint
 /**
  * An end point for the RPC that defines what functions to trigger given a message.
- *
+ *M
```
extra M

andrewor14 commented on core/src/main/scala/org/apache/spark/rpc/RpcEndpoint.scala:
```diff
@@ -44,14 +44,14 @@ private[spark] trait ThreadSafeRpcEndpoint extends RpcEndpoint
  * It is guaranteed that `onStart`, `receive` and `onStop` will be called in sequence.
  *
- * The lift-cycle will be:
+ * The life-cycle of an end point is:
```
endpoint 1 word

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -244,6 +248,10 @@ private[spark] class CoarseMesosSchedulerBackend(
           logDebug(s"Accepting offer: $id with attributes: $offerAttributes mem: $mem cpu: $cpus")
+          if (conf.getBoolean("spark.shuffle.service.enabled", false)) {
+            slaveIdsToShuffleEndpoints(offer.getSlaveId.getValue) =
+              RpcAddress(offer.getHostname, conf.getInt("spark.mesos.shuffle.service.port", 7327))
```
same here, just use old port and config

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -62,6 +64,8 @@ private[spark] class CoarseMesosSchedulerBackend(
 
   val slaveIdsWithExecutors = new HashSet[String]
 
+  val slaveIdsToShuffleEndpoints = new HashMap[String, RpcAddress]
```
this needs a comment: `// Only used if shuffle service is enabled`

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
+            logWarning(s"Unable to connect to shuffle service for slave ${slaveId}" +
+              s", please manually remove shuffle data after driver exit. " +
+              s"Error: ${ref.failed.get}")
+          }
```
these blocks need to be indented

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
+          rpcAddress,
+          MesosExternalShuffleService.ENDPOINT_NAME).onComplete { ref =>
+          if (ref.isSuccess) {
+            ref.get.send(RegisterMesosDriver(conf.getAppId, sc.env.rpcEnv.address))
```
is it safe to just call `get` here? Should we do a null check just in case? Same in L288

andrewor14 replied:
also nit: you use `sc.env.rpcEnv` in a few places, can you make it a local variable `val rpcEnv`

andrewor14 replied:
also, should we do an `askWithRetry` here? Right now this is a one-way message and we have no idea whether this is actually received. More concretely I'm suggesting that you add a `DriverRegistered` response message sent from the shuffle service.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
     logInfo(s"Mesos task $taskId is now $state")
+    val slaveId: String = status.getSlaveId.getValue
     stateLock.synchronized {
+      if (TaskState.fromMesos(state).equals(TaskState.RUNNING) &&
```
this block of code needs a huge comment:
```
// If the shuffle service is enabled, have the driver register with each one
// of them. This is important for cleaning up state on the shuffle services
// after this application exits; there is currently not a great way to detect
// this through Mesos, since the shuffle services are set up independently.
```

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
+    val slaveId: String = status.getSlaveId.getValue
     stateLock.synchronized {
+      if (TaskState.fromMesos(state).equals(TaskState.RUNNING) &&
+        slaveIdsToShuffleEndpoints.contains(slaveId)) {
```
indent 2 spaces

andrewor14 commented on sbin/start-mesos-shuffle-service.sh:
```diff
@@ -0,0 +1,33 @@
+
+# Starts the external shuffle server on the machine this script is executed on.
+#
+# Usage: start-shuffle-server.sh
```
outdated, the comment up there is also outdated

andrewor14 replied:
also, please add in the comment here that "this shuffle service detects when an application exits and automatically cleans up its shuffle files", and the fact that this is hard to do through Mesos.

tnachen replied:
I added a TODO in the CoarseMesosSchedulerBackend

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
+        val rpcAddress = slaveIdsToShuffleEndpoints.remove(slaveId).get
+        logDebug(s"Connecting to shuffle service on slave ${slaveId}, " +
+            s"address $rpcAddress for app ${conf.getAppId}")
+        sc.env.rpcEnv.asyncSetupEndpointRef(
```
why async? just wondering

tnachen replied:
Didn't feel like this needs to be blocking since we need the results, so just ended up doing a async call instead.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    override def onDisconnected(remoteAddress: RpcAddress): Unit = {
+      logDebug(s"Received disconnect from $remoteAddress")
```
this should be at least `logInfo`. Also I would rephrase this:
```
s"Application $appId disconnected (address was $remoteAddress)"
```

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,127 @@
+    // Stores a map of driver rpc addresses to app ids
+    private val connectedApps = new mutable.HashMap[RpcAddress, String]
+
+    override def receive: PartialFunction[Any, Unit] = {
```
as mentioned below, this would become `receiveAndReply`

andrewor14 commented:
@tnachen looks great. My comments are mostly minor but I think this is very close.

SparkQA commented:
[Test build #39344 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39344/console) for   PR 7820 at commit [`1910bd5`](https://github.com/apache/spark/commit/1910bd526c5378ea2a6fb5ff008751ed544a24d3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class RegisterMesosDriver(appId: String, address: RpcAddress)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

tnachen committed:
Address comments.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39419 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39419/consoleFull) for   PR 7820 at commit [`e0f963b`](https://github.com/apache/spark/commit/e0f963b7dce3c20bada2d5b5b8b563a310edca82).

SparkQA commented:
[Test build #39419 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39419/console) for   PR 7820 at commit [`e0f963b`](https://github.com/apache/spark/commit/e0f963b7dce3c20bada2d5b5b8b563a310edca82).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ExternalShuffleService(`
  * `public class JavaKMeansExample `
  * `public class MesosExternalShuffleClient extends ExternalShuffleClient `
  * `public class RegisterDriver extends BlockTransferMessage `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39421 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39421/consoleFull) for   PR 7820 at commit [`aca93da`](https://github.com/apache/spark/commit/aca93da126ca81aa0306d4bfd1e9e97b60f10a53).

SparkQA commented:
[Test build #39421 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39421/console) for   PR 7820 at commit [`aca93da`](https://github.com/apache/spark/commit/aca93da126ca81aa0306d4bfd1e9e97b60f10a53).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ExternalShuffleService(`
  * `public class JavaKMeansExample `
  * `public class MesosExternalShuffleClient extends ExternalShuffleClient `
  * `public class RegisterDriver extends BlockTransferMessage `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39434 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39434/consoleFull) for   PR 7820 at commit [`fadff89`](https://github.com/apache/spark/commit/fadff896194886854c0bae6657a4897f9a48ce21).

SparkQA commented:
[Test build #39434 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39434/console) for   PR 7820 at commit [`fadff89`](https://github.com/apache/spark/commit/fadff896194886854c0bae6657a4897f9a48ce21).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`
  * `class ExternalShuffleService(`
  * `public class JavaKMeansExample `
  * `public class MesosExternalShuffleClient extends ExternalShuffleClient `
  * `public class RegisterDriver extends BlockTransferMessage `
  * `      class SpecificSafeProjection extends $`
  * `case class FromUTCTimestamp(left: Expression, right: Expression)`
  * `case class ToUTCTimestamp(left: Expression, right: Expression)`
  * `case class DateDiff(endDate: Expression, startDate: Expression)`
  * `case class InitCap(child: Expression) extends UnaryExpression with ImplicitCastInputTypes `

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala:
```diff
@@ -37,15 +39,17 @@ import org.apache.spark.util.Utils
+class ExternalShuffleService(
+    sparkConf: SparkConf,
+    securityManager: SecurityManager,
+    transportConf: TransportConf,
```
You don't need to pass this in here. In fact, do we even use this, since we create another one from `sparkConf` down there?

andrewor14 replied:
This interface is confusing: we pass in both a `SparkConf` and a `TransportConf`, even though one is created from another one.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala:
```diff
@@ -37,15 +39,17 @@ import org.apache.spark.util.Utils
+    sparkConf: SparkConf,
+    securityManager: SecurityManager,
+    transportConf: TransportConf,
+    blockHandler: ExternalShuffleBlockHandler)
```
I would do this a different way, just define a
```
protected def newBlockHandler(): ExternalShuffleBlockHandler = {
  new ExternalShuffleBlockHandler(transportConf)
}
```
and then have the subclass override it

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,142 @@
+/**
+ * MesosExternalShuffleServiceEndpoint is a RPC endpoint that receives
+ * registration requests from Spark drivers launched with Mesos.
+ * It detects driver termination and calls the cleanup callback to [[ExternalShuffleService]]
```
nit: just start with `An RPC endpoint that...`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,142 @@
+
+
+/**
+ * MesosExternalShuffleServiceEndpoint is a RPC endpoint that receives
```
nit: just start with `An RPC endpoint that...`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,142 @@
+          if (!existingAppId.equals(appId)) {
+            logError(s"A new app id $appId has connected to existing address $address" +
+              s", removing registered app $existingAppId")
+            applicationRemoved(existingAppId, true)
```
can you name this boolean parameter

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,142 @@
+/**
+ * An extractor object for matching RegisterDriver message.
+ */
+private[mesos] object RegisterDriverParam {
```
this could be strictly `private`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,142 @@
+/**
+ * MesosExternalShuffleService wraps [[ExternalShuffleService]] which provides an additional
+ * endpoint for drivers to associate with. This allows the shuffle service to detect when
+ * a driver is terminated and can further clean up the cached shuffle data.
```
technically incorrect: shuffle data is not cached. Just say the `clean up the shuffle files`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,142 @@
+
+private[spark] object MesosExternalShuffleService extends Logging {
+  val SYSTEM_NAME = "mesosExternalShuffleService"
+  val ENDPOINT_NAME = "mesosExternalShuffleServiceEndpoint"
```
not used, please clean up the code after an update

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosExternalShuffleService.scala:
```diff
@@ -0,0 +1,142 @@
+    conf, securityManager, transportConf, new MesosExternalShuffleBlockHandler(transportConf)) {
+}
+
+private[spark] object MesosExternalShuffleService extends Logging {
```
this whole thing is duplicated from `ExternalShuffleService`. We should at least attempt to reuse some of the code... I think if you make this object inherit the `ExternalShuffleService` object we can get a lot done.

andrewor14 replied:
never mind you can't extend objects.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -17,28 +17,26 @@
+import scala.collection.mutable.{HashMap, HashSet, LinkedHashMap}
+import scala.concurrent.ExecutionContext
+import scala.util.Random
+import scala.util.control.NonFatal
```
these imports should be organized

andrewor14 replied:
actually this file doesn't need to change if we don't modify the constructor as suggested above

andrewor14 commented on network/shuffle/src/main/java/org/apache/spark/network/shuffle/mesos/MesosExternalShuffleClient.java:
```diff
@@ -0,0 +1,63 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MesosExternalShuffleClient extends ExternalShuffleClient {
```
this needs a huge java doc

andrewor14 commented:
@tnachen Thanks for addressing the comments quickly. The latest changes need a bit of clean up, however. Have you had a chance to test the latest changes on a real cluster and verify that the shuffle files did get cleaned up?

tnachen commented:
Yes I tested with a single master and slave and it worked.
I can get to the comments after 7 today.


> On Aug 2, 2015, at 4:12 PM, andrewor14 <notifications@github.com> wrote:
> 
> @tnachen Thanks for addressing the comments quickly. The latest changes need a bit of clean up, however. Have you had a chance to test the latest changes on a real cluster and verify that the shuffle files did get cleaned up?
> 
> —
> Reply to this email directly or view it on GitHub.
>

andrewor14 commented:
Hey @tnachen I opened #7881 after addressing the comments myself. Could you test out the latest changes in that patch instead?

andrewor14 commented:
Can you close this one now that #7881 is already merged?

tnachen commented:
This is already merged.

dragos commented on sbin/start-mesos-shuffle-service.sh:
```diff
@@ -0,0 +1,35 @@
+. "$sbin/spark-config.sh"
+. "$SPARK_PREFIX/bin/load-spark-env.sh"
+
+exec "$sbin"/spark-daemon.sh start org.apache.spark.deploy.mesos.MesosExternalShuffleService 1
```
Do we still need the old one, `start-shuffle-service.sh`? It was only used by Mesos as far as I remember.

andrewor14 replied:
I think we need to keep it for backward compatibility unfortunately.
